rule feature_counts:
    input:
        # list of sam or bam files
        samples = "{sample}.bam",
        annotation = "annotation.gtf",
        # optional input
        #chr_names="",           # implicitly sets the -A flag
        #fasta="genome.fasta"    # implicitly sets the -G flag
    output:
        directory("results/{sample}")
        # multiext(
        #     "results/{sample}",
        #     ".featureCounts",
        #     ".featureCounts.summary",
        #     ".featureCounts.jcounts",
        # ),
    threads: 2
    params:
        strand=0,  # optional; strandness of the library (0: unstranded [default], 1: stranded, and 2: reversely stranded)
        r_path="",  # implicitly sets the --Rpath flag
        extra = "-O -t exon -g gene_id",
        paired = True,
    log:
        "logs/{sample}.log",
    wrapper:
        "file:../../../subread/featurecounts"
